src/analysis/souffle/taint.dl

```diff
@@ -44,15 +44,58 @@
 // is made by `principal`.
 .decl claimHasTag(principal: Principal, accessPath: AccessPath, tag: Tag)
 
+// Manifests can produce base facts of this form, where usually the principal 
+// is a particle (and this fact is entered by a code reviewer with no knowledge 
+// of the taint analysis).
+.decl claimNotEdge(principal: Principal, src: AccessPath, dst: AccessPath)
+
 //-----------------------------------------------------------------------------
 // Rules
 //-----------------------------------------------------------------------------
+
+// This rule is meant to support the functionality of removing edges from 
+// manifests:
+//      - that clearly does not introduce an unsafe escape hatch that could
+//      break policies (because it does not introduce downgrades that would not 
+//      otherwise be possible).
+//      - without data owners needing to know anything about the application 
+//      structure
+//      - without app behavior specification writers needing to know anything 
+//      about the taint analysis or the tags that might appear on a handle
+
+
+// claimNotEdge(p, src, tgt) is an _attempt_ to remove an edge from the 
+// analysis (by principal p)
+// saysDowngrades(p, path, tgt) is an _attempt_ to remove just one tag
+// from a path (by principal p).
+//
+// The rule for downgrades(p, t) in authorization_logic.dl describes when 
+// a downgrade is "safe" and the downgrade really happens (by consulting the
+// owners).
+//
+// Because the analysis propagates the tags on source edges to dest edges, it is safe
+// to remove an edge when ALL of the tags on the source edge can be downgraded.
+// However, we do not want to interfere with other edges that may flow from
+// that source path. Instead, we remove all of the tags on the source from the
+// midpoint access path we add in the middle of each edge. This rule populates a
+// request to downgrade each tag on the edge midpoint.
+//      - it is safe because these requests are only accepted if they would 
+//      normally be by the downgrades(_, _) rule
+//      - it does not require the manifest writer to know anything about the 
+//      tags in the analysis
+//      - data owners do not need to know about paths, and they will grant 
+//      permission for edge removal indirectly by granting permission to remove 
+//      tags (abstract representations of their data).
+
+saysDowngrades(principal, midpoint, tag) :-
```

Comment:
@aferr @markww  Why do we need to associate edges with downgrades at all?

Is this to connect tag ownership to edge-related claims? If so, why do we need to do that? 

Shouldn't there be a completely separate ownership involving edges/access paths?

Reply:
Claims that edges are not present do roughly the same thing as downgrades -- they give a way to "relax" the taint analysis. The downgrade rule for our analysis is designed to allow information release (or similarly, the data to be transformed from sensitive to less-sensitive data) when the owner of the data permits this. It is somewhat similar to the downgrade rules in many systems for decentralized information flow control like [DLM (the orignal)](https://www.cs.cornell.edu/andru/papers/iflow-tosem.pdf) or [Flume(which is a faster read because it is a systems paper)](https://pdos.csail.mit.edu/papers/flume-sosp07.pdf) or many others. 

Because taking edges out of the analysis could inadvertantly allow information leaks that the data owner did not consent to, it is important to make sure that these don't introduce an escape hatch. One relatively clear way to ensure that is to implement the removal of edges in terms of downgrades, which should be possible.

Ideally, it would be nice formalize and prove an information flow theorem that defends against active attacks (i.e., it shows that even if other participants in the system are malicious, there are no "egresses" of data that the owner has not consented to). [Robust declassification](https://www.cs.cornell.edu/andru/papers/csfw01.pdf) is one way of formalizing this for more conventional IFC systems than ours and [this type system](https://www.cis.upenn.edu/~stevez/papers/Zda03.pdf) shows how to generalize it to decentalized label models like DLM. Unfortunately, since we have an authorization mechanism that is fundamentally new (by combining authorization logic with IFC), we would need to also invent a new property ([Opacity](https://www.researchgate.net/publication/220905175_Opacity_Analysis_in_Trust_Management_Systems) also appears related because it involves active attacks in authorization logic).

Even if we don't manage to get to the point where we are proving formal security properties, it would still be nice to give an informal description of this nice property (there are no data transformations / data "egresses" without the consent of the owner) and have this be clearly true from the rules of our system. By introducing a way to remove edges that does not go through this downgrade rule, we probably lose this informal property.